### PR TITLE
feat: Support runes

### PR DESCRIPTION
**Pull request by baseballyama**

### CODE REVIEW

**Comment by baseballyama** [comment]
I could not figure out how to test this because the globals are different for Svelte4 and Svelte5. Once I created the expected files for Svelte5, but there are too many additional lines. Any suggestions?

**Comment by ota-meshi** [comment]
Wow! I noticed this PR after opening #426.  This PR seems to be a little more advanced than #426 as type information for runes is added.

> I could not figure out how to test this because the globals are different for Svelte4 and Svelte5. Once I created the expected files for Svelte5, but there are too many additional lines. Any suggestions?

In #426, the idea is to explicitly opt in to runes mode using an option to make it work.
https://github.com/sveltejs/svelte-eslint-parser/blob/runes-vars/README.md#parseroptionssveltefeaturesrunes

**Comment by baseballyama** [comment]
> In https://github.com/sveltejs/svelte-eslint-parser/pull/426, the idea is to explicitly opt in to runes mode using an option to make it work.

From the user's point of view, using Svelte5 = being able to use runes, so if there is a way to not force users to configure it, that would be good.

About test, how about to extract variables part and prepare only this part for Svelte4 and Svelte5 separately.
Then we can prevent to make duplicated expected files.

https://github.com/sveltejs/svelte-eslint-parser/pull/425/files#diff-8907e5126cad5dd0dd35d0025fdd3086e01a29d3cb72fd9bb5f37b6352ba9048R3-R52

One question, my understanding is that the Svelte compiler is not bundled with `svelte-eslint-parser` so it always uses the version installed by the user, correct?

**Comment by baseballyama** [comment]
I think my todo is there.

- [ ] Reduce duplicated expected files
- [ ] Add tests for runes types
- [ ] Handle `$derived` (I need to understand about this issue at first)

**Comment by baseballyama** [comment]
Oh but your PR supports `.svelte.js` files.
So somehow we need to combine these PRs.

**Comment by ota-meshi** [comment]
> One question, my understanding is that the Svelte compiler is not bundled with svelte-eslint-parser so it always uses the version installed by the user, correct?

Yeah, we specifying `svelte` in `peerDependencies` and `peerDependenciesMeta`, so I expect the user to be able to use whatever version they want to use.

**Comment by baseballyama** [comment]
What do you think is the best way to do Next Action?
- My opinion is that runes should not be configurable and should be reconize automatically by the parser, but since you are more professional in this field than me, please make a decision.
- As for testing, it is possible to eliminate duplicates and I can handle that.
- Regarding support for the `.svelte.js` file, would it be better to merge your PR after this PR is merged? Or we can collaborate on this branch and put them together as one PR, but I have no opinion here.

**Comment by ota-meshi** [comment]
> My opinion is that runes should not be configurable and should be reconize automatically by the parser, but since you are more professional in this field than me, please make a decision.

I think we can consider what our options are after Svelte v5 is released. I believe we can remove options or even change the defaults for options.
However, it currently needs to be marked as an experimental feature, so I think it makes sense to provide it as an option for now.

> would it be better to merge your PR after this PR is merged?

Yeah, I would like to change and merge #426 after this PR is merged.

**Comment by ota-meshi** [comment]
> However, it currently needs to be marked as an experimental feature, so I think it makes sense to provide it as an option for now.

Oh, I think it's a good idea to judge by svelte version rather than using options. Installing svelte v5 means the user is explicitly using pre-release.

https://github.com/sveltejs/svelte-eslint-parser/pull/425/files#diff-cc6f76d59c9667a8e87dc3efb2b017199c77faed21c60201fc2633a0725c8217R15

File: src/parser/globals.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+const globalsForSvelte4: Readonly<string[]> = [
+  "$$slots",
+  "$$props",
+  "$$restProps",
+] as const;
+export const globalsForSvelte5 = [
+  "$state",
+  "$derived",
+  "$effect",
+  "$effect.pre",
```
Comment (ota-meshi) [inline src/parser/globals.ts:12]: Since `$effect.pre` is a member of `$effect`, I don't think it is necessary to define it here.

```suggestion
```

**Review by ota-meshi — CHANGES_REQUESTED** [review]

**Comment by baseballyama** [comment]
Tomorrow I will add more tests for runes.

File: src/parser/globals.ts (hunk 1 above)
Comment (baseballyama) [reply src/parser/globals.ts]: Fixed!

File: src/parser/typescript/analyze/index.ts [hunk 2]
```diff
@@ -215,6 +215,76 @@ function analyzeDollarDollarVariables(
...
+
+  function addSvelte5Globals() {
+    for (const svelte5Global of globalsForSvelte5) {
+      switch (svelte5Global) {
+        case "$state": {
+          if (
+            scopeManager.globalScope!.through.some(
+              (reference) => reference.identifier.name === svelte5Global,
+            )
+          ) {
```
Comment (ota-meshi) [inline src/parser/typescript/analyze/index.ts:228]: By moving this judgment before the `switch`, I think we can eliminate the four same implementations.

**Review by ota-meshi — CHANGES_REQUESTED** [review]
It looks mostly good, but I did make one comment 😃

File: src/parser/typescript/analyze/index.ts (hunk 2 above)
Comment (baseballyama) [reply src/parser/typescript/analyze/index.ts]: Ohhhhh...
https://github.com/sveltejs/svelte-eslint-parser/pull/425/commits/0a61e915bfe1ebb9f63cc658e8fa0720cb482b3c

**Review by ota-meshi — APPROVED** [review]
LGTM!

**Pull request merged**